File: sdks/js/src/responses.ts
import type { Client, Schemas } from "./client.js";
import { ConfigurationError } from "./errors.js";

/** How many items are read per request while unwinding. */
const ITEM_PAGE = 200;

/** An image handed to the model alongside the question. */
export type ImageSource = Schemas["ImageSource"];

export interface CreateResponseOptions {
  images?: ImageSource[];
}

/**
 * The things one turn was made of, in the order they happened.
 *
 * Read rather than watched: this is what the backend wrote down, so it is the same whether
 * the conversation is still going or ended last week. Deltas are not here — a hundred
 * fragments of one sentence are the sentence — so a caller who wants to watch words arrive
 * reads `session.events()` and a caller who wants the shape of a turn reads this.
 */
export class Items {
  private readonly client: Client;
  private readonly sessionId: string;
  /** Empty for every turn in the session, set for one response's own items. */
  private readonly responseId: string;

  constructor(client: Client, sessionId: string, responseId = "") {
    this.client = client;
    this.sessionId = sessionId;
    this.responseId = responseId;
  }

  /**
   * Yields every item, oldest first, fetching a page at a time.
   *
   * Paging is inside rather than outside because a conversation's length is not something
   * the caller chose: `for await (const item of session.responses.items.unwind())` reads a
   * turn or a thousand the same way.
   */
  async *unwind(options: { limit?: number } = {}): AsyncGenerator<Schemas["AgentResponseItem"]> {
    const page = Math.min(options.limit ?? ITEM_PAGE, 1000);
    let offset = 0;
    while (true) {
      const items = await this.list({ limit: page, offset });
      for (const item of items) {
        yield item;
      }
      // A short page is the last page. Asking again to see an empty one would double the
      // requests for every conversation that happens to be a multiple of the page size,
      // which is not worth avoiding one extra round trip in the rare exact-fit case.
      if (items.length < page) {
        return;
      }
      offset += items.length;
    }
  }

  /** One page of items, for a caller doing its own paging. */
  list(options: { limit?: number; offset?: number } = {}): Promise<readonly Schemas["AgentResponseItem"][]> {
    return this.client.get("/v1/agents/sessions/{id}/responses/items", {
      path: { id: this.sessionId },
      query: {
        ...(this.responseId ? { response_id: this.responseId } : {}),
        limit: options.limit,
        offset: options.offset,
      },
    });
  }

  /** Everything in one array, for a conversation short enough to hold. */
  async all(): Promise<readonly Schemas["AgentResponseItem"][]> {
    const collected: Schemas["AgentResponseItem"][] = [];
    for await (const item of this.unwind()) {
      collected.push(item);
    }
    return collected;
  }
}

/**
 * One turn, and a way to read what it was made of.
 *
 * `create` returns as soon as the agent has started answering rather than when it has
 * finished, because a model takes seconds and a request that waited them out would time out
 * on anything worth asking. So this is a handle on an answer in progress: `items.unwind()`
 * reads what has been written down so far, and `session.events()` is what watches it arrive.
 */
export class AgentResponse {
  readonly created: Schemas["AgentResponse"];
  readonly items: Items;

  constructor(client: Client, created: Schemas["AgentResponse"]) {
    this.created = created;
    this.items = new Items(client, created.session_id, created.id);
  }

  /** The backend's id for this turn, empty for a session that records nothing. */
  get id(): string {
    return this.created.id;
  }

  get status(): Schemas["AgentResponse"]["status"] {
    return this.created.status;
  }
}

/**
 * A session's turns.
 *
 * `items` here is the whole conversation flattened, which is how a conversation reads and
 * how it gets rendered: the question, what the agent did about it, what it said, then the
 * next question. A single turn's items come off the handle `create` returns.
 */
export class Responses {
  readonly items: Items;

  private readonly client: Client;
  private readonly sessionId: string;

  constructor(client: Client, sessionId: string) {
    this.client = client;
    this.sessionId = sessionId;
    this.items = new Items(client, sessionId);
  }

  /** Asks the agent something and names the turn it answers as. */
  async create(text: string, options: CreateResponseOptions = {}): Promise<AgentResponse> {
    const created = await this.client.post("/v1/agents/sessions/{id}/responses", {
      path: { id: this.sessionId },
      body: { text, ...(options.images?.length ? { images: options.images } : {}) },
    });
    return new AgentResponse(this.client, created);
  }

  /**
   * Goes back to a response and carries on from there, as though nothing after it was said.
   *
   * The model forgets the later turns and they drop out of `list` and `items`. An item
   * stands for the response it belongs to, so the thing a transcript renders is enough to
   * rewind to. A conversation kept in Stream Chat cannot be rewound, because the channel
   * would still hold the later turns: fork it at the response instead.
   */
  async rewind(
    to: AgentResponse | Schemas["AgentResponse"] | Schemas["AgentResponseItem"] | string,
  ): Promise<void> {
    const responseId =
      typeof to === "string" ? to : "response_id" in to ? to.response_id : to.id;
    if (!responseId) {
      throw new ConfigurationError("a response that was never recorded cannot be rewound to");
    }
    await this.client.post("/v1/agents/sessions/{id}/rewind", {
      path: { id: this.sessionId },
      body: { response_id: responseId },
    });
  }

  /** The turns so far, oldest first. */
  list(options: { limit?: number; offset?: number } = {}): Promise<readonly Schemas["AgentResponse"][]> {
    return this.client.get("/v1/agents/sessions/{id}/responses", {
      path: { id: this.sessionId },
      query: { limit: options.limit, offset: options.offset },
    });
  }
}

File: sdks/rust/src/responses.rs
use crate::client::Client;
use crate::error::{Error, Result};
use crate::operations::{ListResponseItemsQuery, ListResponsesQuery};
use crate::types;

/// How many items are read per request while reading a whole conversation.
const ITEM_PAGE: i64 = 200;

/// A session's turns.
///
/// `items` here is the whole conversation flattened, which is how a conversation reads: the
/// question, what the agent did about it, what it said, then the next question. One turn's
/// items come off the handle `create` returns.
#[derive(Debug, Clone)]
pub struct Responses {
    client: Client,
    session_id: String,
    pub items: Items,
}

impl Responses {
    pub(crate) fn new(client: Client, session_id: &str) -> Self {
        Responses {
            items: Items::new(client.clone(), session_id, ""),
            client,
            session_id: session_id.into(),
        }
    }

    /// Asks the agent something and names the turn it answers as.
    ///
    /// It returns as soon as the agent has started answering, not when it has finished: a
    /// model takes seconds, and the session's events are what watch it arrive.
    pub async fn create(&self, text: &str) -> Result<AgentResponse> {
        self.create_with(&types::CreateResponseRequest {
            text: text.into(),
            images: None,
        })
        .await
    }

    /// [`Responses::create`], with images or anything else the request carries.
    pub async fn create_with(
        &self,
        request: &types::CreateResponseRequest,
    ) -> Result<AgentResponse> {
        let created = self
            .client
            .create_response(&self.session_id, request)
            .await?;
        Ok(AgentResponse {
            items: Items::new(self.client.clone(), &created.session_id, &created.id),
            created,
        })
    }

    /// The turns so far, oldest first.
    pub async fn list(
        &self,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<types::AgentResponse>> {
        self.client
            .list_responses(&self.session_id, &ListResponsesQuery { limit, offset })
            .await
    }

    /// Goes back to a response and carries on from there, as though nothing after it was said.
    ///
    /// The model forgets the later turns and they drop out of `list` and `items`. A
    /// conversation kept in Stream Chat is refused, because the channel would still hold the
    /// later turns: fork it at the response instead.
    pub async fn rewind(&self, to: impl ResponseRef) -> Result<()> {
        let response_id = to.response_id();
        if response_id.is_empty() {
            return Err(Error::configuration(
                "a response that was never recorded cannot be rewound to",
            ));
        }
        self.client
            .rewind_session(
                &self.session_id,
                &types::RewindSessionRequest {
                    response_id: response_id.into(),
                },
            )
            .await
    }
}

/// One turn, and a way to read what it was made of.
#[derive(Debug, Clone)]
pub struct AgentResponse {
    pub created: types::AgentResponse,
    pub items: Items,
}

impl AgentResponse {
    /// The backend's id for this turn. Not the `turn_id` socket events carry.
    pub fn id(&self) -> &str {
        &self.created.id
    }
}

/// The things a conversation, or one turn of it, was made of, in the order they happened.
///
/// Read rather than watched: this is what the backend wrote down. Deltas are not here; a
/// hundred fragments of one sentence are the sentence.
#[derive(Debug, Clone)]
pub struct Items {
    client: Client,
    session_id: String,
    response_id: String,
}

impl Items {
    fn new(client: Client, session_id: &str, response_id: &str) -> Self {
        Items {
            client,
            session_id: session_id.into(),
            response_id: response_id.into(),
        }
    }

    /// One page, for a caller doing its own paging.
    pub async fn list(
        &self,
        limit: Option<i64>,
        offset: Option<i64>,
    ) -> Result<Vec<types::AgentResponseItem>> {
        let query = ListResponseItemsQuery {
            response_id: (!self.response_id.is_empty()).then(|| self.response_id.clone()),
            limit,
            offset,
        };
        self.client
            .list_response_items(&self.session_id, &query)
            .await
    }

    /// Every item, oldest first, a page at a time.
    pub async fn all(&self) -> Result<Vec<types::AgentResponseItem>> {
        let mut collected = Vec::new();
        loop {
            let page = self
                .list(Some(ITEM_PAGE), Some(collected.len() as i64))
                .await?;
            let last = (page.len() as i64) < ITEM_PAGE;
            collected.extend(page);
            // A short page is the last page.
            if last {
                return Ok(collected);
            }
        }
    }
}

/// Anything that names a response: its id, the response, or any item of one.
pub trait ResponseRef {
    fn response_id(&self) -> &str;
}

impl ResponseRef for &str {
    fn response_id(&self) -> &str {
        self
    }
}

impl ResponseRef for &String {
    fn response_id(&self) -> &str {
        self
    }
}

impl ResponseRef for &AgentResponse {
    fn response_id(&self) -> &str {
        &self.created.id
    }
}

impl ResponseRef for &types::AgentResponse {
    fn response_id(&self) -> &str {
        &self.id
    }
}

impl ResponseRef for &types::AgentResponseItem {
    fn response_id(&self) -> &str {
        &self.response_id
    }
}
